### PR TITLE
⚠️ Move Status.APIEndpoints to Spec.ControlPlaneEndpoint

### LINKED ISSUE #1687: RFE: Control plane support for various ways to provide a stable API endpoint
**User Story**

As an operator, I would like to choose my own way to provide a stable API endpoint, as required by the [default control plane management implementation](#1613 ).

**Detailed Description**

The default control plane management implementation requires a stable API endpoint. This endpoint
- MUST route API requests to a live kube-apiserver replica, if one is available. 
- MUST be defined before the first control plane replica is created, so that it is added to the SubjectAltNames of the kube-apiserver server certificate. 
- MUST stay the same for the entire life of the cluster.
- MUST be a DNS record, or an IP.

For reference: Today, the stable API endpoint is provided either by the operator (e.g. using an existing LB, or a VIP backed by keepalived), or by the infra provider (e.g. an ELB created by CAPA). There is a proposal to make an explicit provider for this endpoint: #1250. 

/cc @detiber @randomvariable

### PR DESCRIPTION
**Pull request by vincepri**
Signed-off-by: Vince Prignano <vincepri@vmware.com>

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1687 

/assign @ncdc 

/cc @moshloop @akutz @CecileRobertMichon @noamran @randomvariable @andrewsykim

### CODE REVIEW

**Comment by ncdc** [comment]
@yastij

File: controllers/cluster_controller_phases.go [hunk 1]
```diff
@@ -49,8 +49,8 @@ func (r *ClusterReconciler) reconcilePhase(_ context.Context, cluster *clusterv1
 		cluster.Status.SetTypedPhase(clusterv1.ClusterPhaseProvisioning)
 	}
 
-	// Set the phase to "provisioned" if the infrastructure is ready and APIEndpoints exists
-	if cluster.Status.InfrastructureReady && len(cluster.Status.APIEndpoints) > 0 {
+	// Set the phase to "provisioned" if the infrastructure is ready and ControlPlaneEndpoint is not zero.
```
Comment (ncdc) [inline controllers/cluster_controller_phases.go]: I'm not sure we need this comment any more 😛

File: controllers/cluster_controller_phases.go [hunk 2]
```diff
@@ -174,12 +174,12 @@ func (r *ClusterReconciler) reconcileInfrastructure(ctx context.Context, cluster
...
-		if err := util.UnstructuredUnmarshalField(infraConfig, &cluster.Status.APIEndpoints, "status", "apiEndpoints"); err != nil {
-			return errors.Wrapf(err, "failed to retrieve Status.APIEndpoints from infrastructure provider for Cluster %q in namespace %q",
+	if cluster.Spec.ControlPlaneEndpoint.IsZero() {
+		if err := util.UnstructuredUnmarshalField(infraConfig, &cluster.Spec.ControlPlaneEndpoint, "spec", "controlPlaneEndpoint"); err != nil {
+			return errors.Wrapf(err, "failed to retrieve Spec.ControlPlaneEndpoint from infrastructure provider for Cluster %q in namespace %q",
 				cluster.Name, cluster.Namespace)
-		} else if len(cluster.Status.APIEndpoints) == 0 {
-			return errors.Wrapf(err, "retrieved empty Status.APIEndpoints from infrastructure provider for Cluster %q in namespace %q",
+		} else if cluster.Spec.ControlPlaneEndpoint.IsZero() {
+			return errors.Wrapf(err, "retrieved empty Spec.ControlPlaneEndpoint from infrastructure provider for Cluster %q in namespace %q",
```
Comment (ncdc) [inline controllers/cluster_controller_phases.go]: Do we really want to return an error if the referenced infra spec.controlPlaneEndpoint is unset? Could we return `nil` instead?

File: controllers/cluster_controller_phases.go (hunk 1 above)
Comment (vincepri) [reply controllers/cluster_controller_phases.go]: Fixed!

File: controllers/cluster_controller_phases.go (hunk 2 above)
Comment (vincepri) [reply controllers/cluster_controller_phases.go]: Removed, this only made sense before given that the field was a list

File: docs/book/src/providers/v1alpha2-to-v1alpha3.md [hunk 3]
```diff
@@ -94,7 +94,13 @@
...
+- As a follow up to the changes mentioned above - for the `external.FailuresFrom` function to retain its functionality, external objects
 (e.g., AWSCluster, AWSMachine, etc.) will need to rename the fields as well.
 -  `Status.ErrorReason` should be renamed to `Status.FailureReason`
--  `Status.ErrorMessage` should be renamed to `Status.FailureMessage` 
+-  `Status.ErrorMessage` should be renamed to `Status.FailureMessage`
+
+## The field `Cluster.Status.APIEndpoints` is removed in favor of `Cluster.Spec.ControlPlaneEndpoint`
+
+- The slice in Cluster.Status has been removed and replaced by a single APIEndpoint field under Spec.
+- Infrastructure provider MUST expose their ControlPlaneEndpoint under `Cluster.Spec.ControlPlaneEndpoint` and optionally remove the APIEndpoints field from their InfraCluster.Status struct.
```
Comment (ncdc) [inline docs/book/src/providers/v1alpha2-to-v1alpha3.md]:
```suggestion
- Infrastructure providers MUST expose a ControlPlaneEndpoint field in their cluster infrastructure resource at `Spec.ControlPlaneEndpoint`. They may optionally remove the `Status.APIEndpoints` field (Cluster API no longer uses it).
```

**Comment by ncdc** [comment]
Probably should also update docs/book/src/architecture/controllers/cluster.md as it mentions apiEndpoints

**Pull request merged**